indexer/indexer.go

```diff
@@ -412,69 +446,48 @@ func (ix *Indexer) FetchAndIndexRepo(ctx context.Context, job *crawlWork) error
 		return fmt.Errorf("failed to get repo root: %w", err)
 	}
 
+	// attempt to process buffered events
 	if !job.initScrape && len(job.catchup) > 0 {
 		first := job.catchup[0]
+		var resync bool
 		if first.evt.Since == nil || rev == *first.evt.Since {
-			for _, j := range job.catchup {
+			for i, j := range job.catchup {
 				catchupEventsProcessed.Inc()
 				if err := ix.repomgr.HandleExternalUserEvent(ctx, pds.ID, ai.Uid, ai.Did, j.evt.Since, j.evt.Rev, j.evt.Blocks, j.evt.Ops); err != nil {
-					// TODO: if we fail here, we should probably fall back to a repo re-sync
-					return fmt.Errorf("post rebase catchup failed: %w", err)
+					log.Errorw("buffered event catchup failed", "error", err, "did", ai.Did, "i", i, "jobCount", len(job.catchup))
+					resync = true // fall back to a repo sync
+					break
 				}
 			}
 
-			return nil
+			if !resync {
+				return nil
+			}
 		}
 	}
 
-	var host string
-	if pds.SSL {
-		host = "https://" + pds.Host
-	} else {
-		host = "http://" + pds.Host
-	}
-	c := &xrpc.Client{
-		Host: host,
-	}
-
-	ix.ApplyPDSClientSettings(c)
-
 	if rev == "" {
 		span.SetAttributes(attribute.Bool("full", true))
 	}
 
-	limiter := ix.GetLimiter(pds.ID)
-	if limiter == nil {
-		limiter = rate.NewLimiter(rate.Limit(pds.CrawlRateLimit), 1)
-		ix.SetLimiter(pds.ID, limiter)
-	}
-
-	// Wait to prevent DOSing the PDS when connecting to a new stream with lots of active repos
-	limiter.Wait(ctx)
+	c := models.ClientForPds(&pds)
+	ix.ApplyPDSClientSettings(c)
 
-	log.Infow("SyncGetRepo", "did", ai.Did, "user", ai.Handle, "since", rev)
-	// TODO: max size on these? A malicious PDS could just send us a petabyte sized repo here and kill us
-	repo, err := comatproto.SyncGetRepo(ctx, c, ai.Did, rev)
+	repo, err := ix.fetchRepo(ctx, c, &pds, ai.Did, rev)
 	if err != nil {
-		reposFetched.WithLabelValues("fail").Inc()
-		return fmt.Errorf("failed to fetch repo: %w", err)
+		return err
 	}
-	reposFetched.WithLabelValues("success").Inc()
 
 	// this process will send individual indexing events back to the indexer, doing a 'fast forward' of the users entire history
 	// we probably want alternative ways of doing this for 'very large' or 'very old' repos, but this works for now
 	if err := ix.repomgr.ImportNewRepo(ctx, ai.Uid, ai.Did, bytes.NewReader(repo), &rev); err != nil {
 		span.RecordError(err)
 
 		if ipld.IsNotFound(err) {
```

Comment:
Do we want to log when this happens? Also can we make the error messages a bit more informative i.e. `failed to full-resync repo after catchup failure: %w` or something?

Reply:
yeah, thats a good idea